nature/hmac.py

```diff
@@ -54,40 +47,45 @@ def is_valid(self):
         :return: True if the request is valid
         :rtype: bool
         """
-        if self._is_valid is None:
-            self._is_valid = self.within_clock_skew and self.has_valid_credentials
-        return self._is_valid
+        return self.within_clock_skew and self.has_valid_credentials
 
     @property
     def has_admin_group(self):
         """Return true if the user has an admin group"""
-        return self.is_valid and bool(set(self.ADMIN_GROUPS).intersection(self.groups))
+        return any([group.is_admin_group for group in self.groups])
 
     @property
     def has_office_hki_group(self):
         """Return true if user has an office hki group"""
-        return self.is_valid and bool(
-            set(self.OFFICE_HKI_GROUPS).intersection(self.groups)
-        )
+        return any([group.is_office_hki_group for group in self.groups])
 
     @property
     def has_office_group(self):
         """Return true if user has an office group"""
-        return self.is_valid and bool(set(self.OFFICE_GROUPS).intersection(self.groups))
+        return any([group.is_office_group for group in self.groups])
 
     @property
     def groups(self):
-        return self.request.META.get("HTTP_X_FORWARDED_GROUPS", "").split(";")
+        group_names = self.request.META.get("HTTP_X_FORWARDED_GROUPS", "").split(";")
+        return HMACGroup.objects.filter(name__in=group_names)
```

Comment:
This works just fine in this case as you are still just iterating over the groups in the previous functions. I did not check the entire code base here but just wanted to make sure you have taken into consideration that this now returns a queryset and not a list anymore.

Reply:
Yes, that's intended change as we are now saving the groups in the database